javascript/stimulus_reflex.js

```diff
@@ -1,166 +1,75 @@
-              controllerElement,
-              reflexId
-            )
-            registerReflex(reflexData)
-          }
-
-          if (reflexes[reflexId]) {
-            reflexes[reflexId].totalOperations = totalOperations
-            reflexes[reflexId].pendingOperations = totalOperations
-            reflexes[reflexId].completedOperations = 0
-            CableReady.perform(reflexOperations)
-          }
-        }
+class StimulusReflexController extends Controller {
+  constructor (...args) {
+    super(...args)
+    register(this)
+  }
+}
 
-        // run piggy back operations after stimulus reflex behavior
-        CableReady.perform(data.operations)
-      },
-      connected: () => {
-        actionCableSubscriptionActive = true
-        emitEvent('stimulus-reflex:connected')
-      },
-      rejected: () => {
-        actionCableSubscriptionActive = false
-        emitEvent('stimulus-reflex:rejected')
-        if (Debug.enabled) console.warn('Channel subscription was rejected.')
-      },
-      disconnected: willAttemptReconnect => {
-        actionCableSubscriptionActive = false
-        emitEvent('stimulus-reflex:disconnected', willAttemptReconnect)
-      }
-    })
+// Initializes StimulusReflex by registering the default Stimulus controller with the passed Stimulus application.
+//
+// - application - the Stimulus application
+// - options
+//   * controller - [optional] the default StimulusReflexController
+//   * consumer - [optional] the ActionCable consumer
+//   * debug - [false] log all Reflexes to the console
+//   * params - [{}] key/value parameters to send during channel subscription
+//   * isolate - [false] restrict Reflex playback to the tab which initiated it
+//
+const initialize = (application, initializeOptions = {}) => {
+  const { controller, consumer, debug, params, isolate } = initializeOptions
+  actionCable.consumer = consumer
+  actionCable.params = params
+  isolationMode.set(!!isolate)
+  reflexes.app = application
+  reflexes.app.schema = { ...defaultSchema, ...application.schema }
+  reflexes.app.register(
+    'stimulus-reflex',
+    controller || StimulusReflexController
+  )
+  Debug.set(!!debug)
+  const observer = new MutationObserver(setupDeclarativeReflexes)
+  observer.observe(document.documentElement, {
+    attributeFilter: [reflexes.app.schema.reflexAttribute],
+    childList: true,
+    subtree: true
+  })
 }
 
-// Extends a regular Stimulus controller with StimulusReflex behavior.
+// Registers a Stimulus controller and extends it with StimulusReflex behavior
 //
-// Methods added to the Stimulus controller:
-// - stimulate
-// - __perform
+// controller - the Stimulus controller
+// options - [optional] configuration
 //
-const extendStimulusController = controller => {
+const register = (controller, options = {}) => {
+  const channel = 'StimulusReflex::Channel'
+  controller.StimulusReflex = { ...options, channel }
+  actionCable.createSubscription(controller)
```

Comment:
May I suggest that you modify register to this.
```
const register = (controller, options = {}, consumer = actionCable) => 
```
then you'd have 
```
consumer.createSubscription(controller)
```

Reply:
Yep, my gut feeling tells me that even `createSubscription` is something very websocket-specific.

my take would be to make `transports/action_cable` an actual adapter `class`, like in @jonathan-s 's original implementation, so that we can later swap it out for a different (AJAX) adapter more easily

Reply:
There will be no `consumer` option in v4; see #438. Consumer will come from the Stimulus application.

As I mentioned in the PR description, I exercised significant restraint in not changing the API. Even there, I did remove a vestigal and antiquated technique we had in place as a fallback for finding consumers because it was itself raising deprecation errors in Chrome. So, save your good ideas for changes until after v3.4.2 comes out!